From dd51b9cbc145aee8c42d711104c1d4085e537485 Mon Sep 17 00:00:00 2001
From: David Sniatkiwicz <davidsn@marvell.com>
Date: Tue, 20 Mar 2018 11:57:09 +0200
Subject: [PATCH 0578/1051] icu: a8kp: remap ICU entries to their GICv3 offset
 in AP810

ICU interrupts in AP806 are written to GIC-P by writing only source ID,
than the GIC-P remaps the interrupt IDs according to AP806 GIC
assignments:
- 0-31 : PPI/SGI
- 32-64: AP interrupts
So the remap was done internally by HW.

AP810 has no GIC-P in between, and the ICU writes the message directly
to
the GIC, so the differences are:
1. AP810 GIC has 92 interrupt entries, pre-assigned to:
   0-31 : PPI/SGI
   32-92: AP interrupts
2. ICU writes message ID directly to GIC, hence it should also increase
   the GIC entries ID by 92, (this has to be done by SW).
3. AP806 interrupts are divided to 2 groups:  64-128 and 288-352,
   while for AP810 they are consecutive, starting at 92.

Change-Id: I319e0b11a5734222cd553845fef3272075e2d01f
Signed-off-by: David Sniatkiwicz <davidsn@marvell.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/irqchip/irq-mvebu-icu.c | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/drivers/irqchip/irq-mvebu-icu.c b/drivers/irqchip/irq-mvebu-icu.c
index 13063339b416..128ae6031ab1 100644
--- a/drivers/irqchip/irq-mvebu-icu.c
+++ b/drivers/irqchip/irq-mvebu-icu.c
@@ -36,6 +36,9 @@
 #define ICU_SATA0_ICU_ID	109
 #define ICU_SATA1_ICU_ID	107
 
+/* AP810 ICU entries from CP110 are consecutive */
+#define ICU_GIC_SPI_AP810_BASE		92
+
 struct mvebu_icu {
 	struct irq_chip irq_chip;
 	void __iomem *base;
@@ -74,6 +77,27 @@ static void mvebu_icu_write_msg(struct msi_desc *desc, struct msi_msg *msg)
 		mvebu_icu_init(icu, msg);
 		/* Configure the ICU with irq number & type */
 		icu_int = msg->data | ICU_INT_ENABLE;
+		/*
+		 * identify if it's a CP110 connected to AP810 or CP110
+		 * connected to AP806, since they differ in relevance to how
+		 * they are connected to AP GIC entries.
+		 * ICU interrupts in AP806 are written to GIC-P,
+		 * by writing only source ID, than the GIC-P remaps these
+		 * interrupt IDs, according to AP806 GIC assignments:
+		 * - 0-31 : PPI/SGI
+		 * - 32-64: AP interrupts
+		 * So the remap was done internally by HW.
+		 *
+		 * For AP810, there is no GIC-P in between, and the ICU writes
+		 * the message directly to the GIC, so the differences are:
+		 * 1. AP810 GIC has 92 interrupt entries, pre-assigned to:
+		 * 0-31 : PPI/SGI
+		 * 32-92: AP interrupts
+		 * 2. ICU need to write message ID directly to GIC, hence it
+		 * should also increase the GIC entries ID by 92, by SW
+		 */
+		if (of_machine_is_compatible("marvell,armada-ap810"))
+			icu_int += ICU_GIC_SPI_AP810_BASE;
 		if (icu_irqd->type & IRQ_TYPE_EDGE_RISING)
 			icu_int |= ICU_IS_EDGE;
 		icu_int |= icu_irqd->icu_group << ICU_GROUP_SHIFT;
-- 
2.17.1

